hedera-mirror-rosetta/app/domain/types/constants_test.go

```diff
@@ -27,9 +27,17 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+var hederaFunctionalityLength = 67
+
 func TestResponseCodeUpToDate(t *testing.T) {
 	// given
 	for code, name := range proto.ResponseCodeEnum_name {
 		assert.Equal(t, name, TransactionResults[code])
 	}
 }
+
+func TestTransactionTypesUpToDate(t *testing.T) {
+	//There isn't a dedicate enum for TransactionBody values, so just check that no new HederaFunctionality exists
+	//Last entry is 80: TokenUnpause
+	assert.Equal(t, len(proto.HederaFunctionality_name), hederaFunctionalityLength)
```

Comment:
this limits us to be only able to change the go constants after sdk release has added the new transaction types.

instead we should do

```suggestion
	assert.LessOrEqual(t, len(proto.HederaFunctionality_name), len(TransactionTypes))
```

Reply:
I think I missed the fact `proto.HederaFunctionality_name` is a superset of the transaction types.

However, the suggestion still stands and there is a way to get the list of `TransactionBody.Data` data cases.

Reply:
tested and it works:

```go
func TestTransactionTypesUpToDate(t *testing.T) {
	sdkTransactionTypes := getSdkTransactionTypes()
	for protoId, name := range sdkTransactionTypes {
		assert.Equal(t, name, TransactionTypes[protoId])
	}
}

func getSdkTransactionTypes() map[int32]string {
	body := proto.TransactionBody{}
	dataFields := body.ProtoReflect().Descriptor().Oneofs().ByName("data").Fields()
	transactionTypes := make(map[int32]string)
	for i := 0; i < dataFields.Len(); i++ {
		dataField := dataFields.Get(i)
		name := strings.ToUpper(string(dataField.Name()))
		transactionTypes[int32(dataField.Number())] = strings.ReplaceAll(name, "_", "")
	}

	return transactionTypes
}
```

Reply:
I don't think that test would work, since `HederaFunctionality` is a superset of `TransactionTypes`, including values such as `NetworkGetExecutionTime` and things that we don't care about.  This test isn't really about `TransactionType`, it's just a warning test that says "Hey, if this fails it means new values were added to `HederaFunctionality`, so there are probably new Transaction Types as well.  I can add a comment saying as much.

Reply:
Didn't refresh the page to see the second comment...  Will test this out.  Anything similar for the Javascript version?

Reply:
the code snippet does exactly what we want and it's stricter. In addition, we don't have the burden of updating the constant `67` in the test file everytime we update the `TransactionTypes` in `constants.go`:

1. `getSdkTransactionTypes` extracts the protoId to transaction type map from the sdk by the reflection mechanism provided by the `protobuf` package.
2. The test case then checks that the `sdkTransactionTypes` is a subset of `TransactionTypes`

So we are allowed to be ahead of SDK, but not the vise versa.

Reply:
> Didn't refresh the page to see the second comment... Will test this out. Anything similar for the Javascript version?

Nah, checked the js code, it's very raw, basically the generated js code directly write the encoded data base value...

Reply:
Cool, I've replaced the Go test then with your test, appreciate the help, that is much cleaner.  Unfortunate that the JS is much more restrictive.